homeassistant/components/mqtt/fan.py

```diff
@@ -217,19 +315,98 @@ def state_received(msg):
                 "qos": self._config[CONF_QOS],
             }
 
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def percentage_received(msg):
+            """Handle new received MQTT message for the percentage."""
+            numeric_val_str = self._templates[ATTR_PERCENTAGE](msg.payload)
+            try:
+                percentage = ranged_value_to_percentage(
+                    self._speed_range, int(numeric_val_str)
+                )
+            except Exception as ex:
+                raise NotValidSpeedError(
+                    f"{msg.payload} is not a valid percentage"
+                ) from ex
+            if 0 <= percentage <= 100:
+                self._percentage = percentage
+            else:
+                raise NotValidSpeedError(f"{msg.payload} is not a valid percentage")
+            self.async_write_ha_state()
+
+        if self._topic[CONF_PERCENTAGE_STATE_TOPIC] is not None:
+            topics[CONF_PERCENTAGE_STATE_TOPIC] = {
+                "topic": self._topic[CONF_PERCENTAGE_STATE_TOPIC],
+                "msg_callback": percentage_received,
+                "qos": self._config[CONF_QOS],
+            }
+            self._percentage = 0
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def preset_mode_received(msg):
+            """Handle new received MQTT message for preset mode."""
+            preset_mode = self._templates[ATTR_PRESET_MODE](msg.payload)
+            if preset_mode in self.preset_modes:
+                self._preset_mode = preset_mode
+                if not self._implemented_percentage and (
+                    preset_mode in self.speed_list
+                ):
+                    self._percentage = ordered_list_item_to_percentage(
```

Comment:
When is a preset mode converted to a percentage?

Reply:
A preset mode is converted when there is no `percentage_command_topic` defined. In that case speeds from the preset_mode list (excluding non speeds) are converted to a percentage. Legacy speeds will will be prepended in the `preset_modes` list when a `speed_command_topic` is set.
When percentages are enabled, the percentage in that case is set through asycn_set_percentage() and the `percentage_state_topic (if not optimistic).

Reply:
https://developers.home-assistant.io/docs/core/entity/fan/#preset-modes

Percentage should probably be set to `None` in this case.

Reply:
If you mean the initial value for self._percentage should be None, I agree. I have changed de code.